src/main/java/com/google/sps/servlets/MultipleTagsServlet.java

```diff
@@ -0,0 +1,115 @@
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.HashMap;
+import org.json.simple.JSONObject;
+import java.util.Set;
+import java.util.HashSet;
+
+/** Servlet that returns the recipeIds of the recipes that have the specified tagnames */
+@WebServlet("/multiple-tags")
+public class MultipleTagsServlet extends HttpServlet {
+  private static final String AUTHORIZATION_ERROR = "User needs to login";
+  
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    JSONObject responseMap = new JSONObject();
+    Gson gson = new Gson();  
+    String json;
+
+    // user's tags selection
+    String[] tagNames = request.getParameterValues("tag-names");
+    if (userService.isUserLoggedIn()) {
+      Query query = getQueryWithFilters(tagNames, userService);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      PreparedQuery results = datastore.prepare(query);
+
+      HashSet<Long> recipeList = new HashSet<>();
+      for (Entity entity : results.asIterable()) {
+        long recipeId = (long) entity.getProperty("recipeId");
+        recipeList.add(recipeId);
+      }
+      responseMap.put("recipeList", recipeList);
+
+    } else {
+      responseMap.put("error", AUTHORIZATION_ERROR); 
+    }
+    
+    json = gson.toJson(responseMap);
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  // helper function for creating query with a combination of filters
+  public Query getQueryWithFilters(String[] tagNames, UserService userService) {
+    List<Query.Filter> filterList = new ArrayList<>();
+    String userId = userService.getCurrentUser().getUserId();
+
+    // set up filters for query over user's tags
+    filterList.add(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userId));
+    
+    // add filters for tagnames
+    List<Query.Filter> tagFilters = new ArrayList<>();
+    if (tagNames != null) {
+      if (tagNames.length > 1) {
+        for (int i = 0; i < tagNames.length; i++) {
+          tagFilters.add(new Query.FilterPredicate("tagName", Query.FilterOperator.EQUAL, tagNames[i]));
+        }
+        filterList.add(new Query.CompositeFilter(Query.CompositeFilterOperator.OR, tagFilters));
+      } else if (tagNames.length == 1) {
+        filterList.add(new Query.FilterPredicate("tagName", Query.FilterOperator.EQUAL, tagNames[0]));
+      }
```

Comment:
This can be simplified to:
for (String tagName : tagNames) {
  tagFilters.add(...);
}
if (tagNames.length > 1) {
  filterList.add(...)
}